File: GeneticAlgorithm/src_swift/GAPopulation.swift
//
//  GAPopulation.swift
//  GeneticAlgorithm_Swift
//
//  Created by Jonathan Sullivan on 7/2/15.
//  Copyright (c) 2015 Jonathan Sullivan. All rights reserved.
//

import Foundation


typealias GAChromosomeString = GAChromosome<String>

let CharacterSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 ~`!@#$%^&*()_-+=[]{}|:,.<>?\""


enum GAPopulationCrossoverType: Int
{
    case GAPopulationCrossoverTypeOnePoint = 0
    case GAPopulationCrossoverTypeTwoPoint
}

enum GAPopulationSelectionType: Int
{
    case GAPopulationSelectionTypeTournament = 0
    case GAPopulationSelectionTypeRouletteWheel
    case GAPopulationSelectionTypeRandom
}


class GAPopulation
{
    
    // MARK: - Private Class Members
    
    private var targetGenes: String
    private var chromosomes: Array<GAChromosomeString>
    
    // MARK: -
    
    
    // MARK: - Private Class Members
    
    var tournamentSize: Int = 2
    var elitismEnabled: Bool = true
    var elitismPercentage: Double = 0.25
    var mutationProbability: Double = 0.25
    var crossoverProbability: Double = 0.95
    var crossoverType: GAPopulationCrossoverType = .GAPopulationCrossoverTypeOnePoint
    var selectionType: GAPopulationSelectionType = .GAPopulationSelectionTypeTournament
    
    // MARK: -
    
    
    // MARK: - Constructors
    
    init(thePopulationSize: Int, theTargetGenes: String)
    {
        self.targetGenes = theTargetGenes
        self.chromosomes = [GAChromosomeString]()
        
        for i in 0..<thePopulationSize
        {
            let randomGenes = self.generateRandomGenes(count(self.targetGenes))
            
            let chromosome = GAChromosomeString(genes: randomGenes)
            chromosome.fitness = self.calculateFitness(chromosome)
            self.chromosomes.append(chromosome)
        }
    }
    
    // MARK: -
    
    
    // MARK: - Helper Methods
    
    private func generateRandomGene() -> Character
    {
        let numberOfCharacters = count(CharacterSet)
        let index = Int(arc4random_uniform(UInt32(numberOfCharacters)))
        
        return CharacterSet[index]
    }
    
    private func generateRandomGenes(geneLength: Int) -> String
    {
        var genes = ""
        
        let characterSetLength = count(CharacterSet)
        
        for _ in 0..<geneLength
        {
            let index = Int(arc4random_uniform(UInt32(characterSetLength)))
            genes += CharacterSet[index]
        }
        
        return genes
    }
    
    // MARK: -
    
    
    // MARK: - Fitness Methods
    
    private func getFittest() -> GAChromosomeString
    {
        var bestChromosome = self.chromosomes[0]
        
        let populationSize = count(self.chromosomes)
        
        for chromosome in self.chromosomes
        {
            if chromosome.fitness < bestChromosome.fitness
            {
                bestChromosome = chromosome
            }
        }
        
        return bestChromosome
    }
    
    private func calculateFitness(chromosome: GAChromosomeString) -> Int
    {
        var fitness = 0
        
        let geneLength = count(chromosome.genes)
        
        for i in 0..<geneLength
        {
            if Character(chromosome.genes[i]) == Character(self.targetGenes[i])
            {
                fitness += 1
            }
        }
        
        let targetGenesLength = count(self.targetGenes)
        return abs(fitness - targetGenesLength)
    }
    
    // MARK: -
    
    
    // MARK: - Mutation Methods
    
    private func mutate(chromosome: GAChromosomeString)
    {
        let targetLength = count(targetGenes)
        let pivot = Int(arc4random_uniform(UInt32(targetLength)))
        
        var genes = Array(chromosome.genes)
        genes[pivot] = self.generateRandomGene()
        
        var mutatedGenes = String()
        mutatedGenes.extend(genes)
        
        chromosome.genes = mutatedGenes
    }
    
    // MARK: -
    
    
    // MARK: - Selection Methods
    
    private func selectChromosome() -> GAChromosomeString
    {
        let chromosome: GAChromosomeString
        
        switch (self.selectionType)
        {
            case .GAPopulationSelectionTypeRouletteWheel:
                chromosome = self.rouletteWheelSelection()
            break
            
            case .GAPopulationSelectionTypeTournament:
                chromosome = self.tournamentSelection(self.tournamentSize)
            break
            
            default:
                chromosome = self.randomSelection()
            break
        }
        
        return chromosome
    }
    
    private func randomSelection() -> GAChromosomeString
    {
        let populationSize = count(self.chromosomes)
        
        let index = Int(arc4random_uniform(UInt32(populationSize)))
        return self.chromosomes[index]
    }
    
    private func rouletteWheelSelection() -> GAChromosomeString
    {
        let populationSize = count(self.chromosomes)
        
        var totalFitness = 0
        
        for chromosome in self.chromosomes
        {
            totalFitness += chromosome.fitness
        }
        
        let randomNumber = Double(Float(arc4random()) / Float(UINT32_MAX))
        var accumulatedNormalizedFitness = 0.0
        
        var chromosome = self.chromosomes[0]
        
        for i in 0..<populationSize
        {
            chromosome = self.chromosomes[i]
            
            accumulatedNormalizedFitness += (Double(chromosome.fitness) / Double(totalFitness))
            
            if randomNumber < accumulatedNormalizedFitness
            {
                break
            }
        }
        
        return chromosome
    }
    
    private func tournamentSelection(tournamentSize: Int) -> GAChromosomeString
    {
        var competitors = [GAChromosomeString]()
        
        let populationSize = count(self.chromosomes)
        
        for i in 0..<tournamentSize
        {
            let index = Int(arc4random_uniform(UInt32(populationSize)))
            competitors.append(self.chromosomes[index])
        }
        
        var strongestCompetitor = competitors[0]
        
        for i in 1..<tournamentSize
        {
            var competitor = self.chromosomes[i]
            
            if competitor.fitness < strongestCompetitor.fitness
            {
                strongestCompetitor = competitor
            }
        }
        
        return strongestCompetitor
    }
    
    // MARK: -
    
    
    // MARK: - Crossover Methods
    
    private func crossoverPopulation()
    {
        var elitismOffset = 0;
        let populationSize = count(self.chromosomes)
        
        var buffer = [GAChromosomeString]();
        
        if (self.elitismEnabled)
        {
            elitismOffset = Int(self.elitismPercentage * Double(populationSize))
            
            self.chromosomes.sort({$0.fitness < $1.fitness})
            
            for i in 0..<elitismOffset
            {
                let chromosome = self.chromosomes[i]
                buffer.append(chromosome)
            }
        }
        
        for var i = elitismOffset; i < populationSize; i += 2
        {
            let chromosomeA = self.selectChromosome()
            let chromosomeB = self.selectChromosome()
            
            let randomCrossover = Double(Float(arc4random()) / Float(UINT32_MAX))
            
            if randomCrossover < self.crossoverProbability
            {
                let offspring: (first: GAChromosomeString, second: GAChromosomeString)
                
                switch (self.crossoverType)
                {
                case .GAPopulationCrossoverTypeTwoPoint:
                    offspring = self.twoPointCrossover(chromosomeA, chromosomeB: chromosomeB)
                    break
                    
                default:
                    offspring = self.onePointCrossover(chromosomeA, chromosomeB: chromosomeB)
                    break
                }
                
                buffer.append(offspring.first)
                buffer.append(offspring.second)
            }
            else {
                buffer.append(chromosomeA)
                buffer.append(chromosomeB)
            }
        }
        
        self.chromosomes = Array(buffer)
    }
    
    private func onePointCrossover(chromosomeA: GAChromosomeString, chromosomeB: GAChromosomeString) -> (GAChromosomeString, GAChromosomeString)
    {
        let targetLength = Int(count(targetGenes))
        let pivot = Int(arc4random_uniform(UInt32(targetLength)))
        
        let genesToCrossOverA = chromosomeA.genes.substring(pivot, end: targetLength)
        let genesToCrossOverB = chromosomeB.genes.substring(pivot, end: targetLength)
        
        let crossedOverGenesA = chromosomeA.genes.substring(0, end: pivot) + genesToCrossOverB
        let crossedOverGenesB = chromosomeB.genes.substring(0, end: pivot) + genesToCrossOverA
        
        let offspringA = GAChromosomeString(genes: crossedOverGenesA)
        let offspringB = GAChromosomeString(genes: crossedOverGenesB)
        
        let randomMutation = Double(Float(arc4random()) / Float(UINT32_MAX))
        
        if randomMutation < self.mutationProbability
        {
            self.mutate(offspringA)
            self.mutate(offspringB)
        }
        
        offspringA.fitness = self.calculateFitness(offspringA)
        offspringB.fitness = self.calculateFitness(offspringB)
        
        return (offspringA, offspringB)
    }
    
    private func twoPointCrossover(chromosomeA: GAChromosomeString, chromosomeB: GAChromosomeString) -> (GAChromosomeString, GAChromosomeString)
    {
        let targetLength = count(self.targetGenes)
        
        var pivotA = Int(arc4random_uniform(UInt32(targetLength)))
        var pivotB = Int(arc4random_uniform(UInt32(targetLength)))
        
        if (pivotA > pivotB)
        {
            let temp = pivotA
            pivotA = pivotB
            pivotB = temp
        }
        
        let genesToCrossOverA = chromosomeA.genes.substring(pivotA, end: pivotB)
        let genesToCrossOverB = chromosomeB.genes.substring(pivotA, end: pivotB)
        
        let crossedOverGenesA = chromosomeA.genes.substring(0, end: pivotA) + genesToCrossOverB + chromosomeA.genes.substring(pivotB, end: targetLength)
        let crossedOverGenesB = chromosomeB.genes.substring(0, end: pivotA) + genesToCrossOverA + chromosomeA.genes.substring(pivotB, end: targetLength)
        
        let offspringA = GAChromosomeString(genes: crossedOverGenesA)
        let offspringB = GAChromosomeString(genes: crossedOverGenesB)
        
        let randomMutation = Double(Float(arc4random()) / Float(UINT32_MAX))
        
        if randomMutation < self.mutationProbability
        {
            self.mutate(offspringA)
            self.mutate(offspringB)
        }
        
        offspringA.fitness = self.calculateFitness(offspringA)
        offspringB.fitness = self.calculateFitness(offspringB)
        
        return (offspringA, offspringB)
    }
    
    // MARK: -
    
    
    // MARK: - Evolution Methods
    
    func evolve()
    {
        var generationCount = 0
        
        while (self.getFittest().fitness != 0)
        {
            self.crossoverPopulation()
            
            print("[Generation \(generationCount)]\n")
            print("Genes: \(self.getFittest().genes)\n")
            print("Fitness: \(self.getFittest().fitness)\n\n")
            
            generationCount++
        }
    }
    
    // MARK: -
}

File: Genetic Algorithm/src_c++/GAPopulation.cpp
//
//  GAPopulation.cpp
//  GeneticAlgorithm_C++
//
//  Created by Jonathan Sullivan on 6/28/15.
//
// The MIT License (MIT)
//
// Copyright (c) 2015 Jonathan Sullivan
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <cstdlib>

#include "GAPopulation.h"
#include "GAChromosome.h"
#include "GAChromosome.cpp"


#pragma mark - Namespaces

using namespace std;

#pragma mark -


#pragma mark - Constants

static const string kCharacterSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 ~`!@#$%^&*()_-+=[]{}|;:,.<>?\"";

#pragma mark -


#pragma mark - Constructors

GAPopulation::GAPopulation(int thePopulationSize, string theTargetGenes)
{
    this->targetGenes = theTargetGenes;
    this->chromosomes = vector<GAChromosomeString>();
    
    for (int i = 0; i < thePopulationSize; i++)
    {
        string randomGenes = this->generateRandomGenes((int)this->targetGenes.length());
        
        GAChromosomeString chromosome = GAChromosomeString(randomGenes);
        chromosome.fitness = this->calculateFitness(chromosome);
        this->chromosomes.push_back(chromosome);
    }
}

#pragma mark -


#pragma mark - Mutation Methods

void GAPopulation::mutate(GAChromosomeString &chromosome)
{
    size_t targetLength = this->targetGenes.length();
    
    int pivot = (rand() % targetLength);
    
    string genes = chromosome.genes;
    genes.at(pivot) = this->generateRandomGene();
    chromosome.genes = genes;
}

#pragma mark -


#pragma mark - Selection Methods

GAChromosomeString GAPopulation::tournamentSelection(int theTournamentSize)
{
    vector<GAChromosomeString> competitors = vector<GAChromosomeString>();
    
    int populationSize = (int)this->chromosomes.size();
    
    for (int i = 0; i < theTournamentSize; i++)
    {
        int index = (rand() % populationSize);
        competitors.push_back(this->chromosomes.at(index));
    }
    
    GAChromosomeString strongestCompetitor = competitors.at(0);
    
    for (int i = 1; i < theTournamentSize; i++)
    {
        GAChromosomeString competitor = this->chromosomes.at(i);
        
        if (competitor.fitness < strongestCompetitor.fitness)
            strongestCompetitor = competitor;
    }
    
    competitors.clear();
    
    return strongestCompetitor;
}

GAChromosomeString GAPopulation::rouletteWheelSelection()
{
    int populationSize = (int)this->chromosomes.size();
    
    int totalFitness = 0;
    
    for (int i = 0; i < populationSize; i++)
        totalFitness += this->chromosomes.at(i).fitness;
    
    double randomNumber = (float)(rand() / (float)RAND_MAX);
    double accumulatedNormalizedFitness = 0;
    
    GAChromosomeString chromosome;
    
    for (int i = 0; i < populationSize; i++)
    {
        GAChromosomeString theChromosome = this->chromosomes.at(i);
        
        accumulatedNormalizedFitness += (theChromosome.fitness / (double)totalFitness);
        
        if (randomNumber < accumulatedNormalizedFitness)
        {
            chromosome = theChromosome;
            break;
        }
    }
    
    return chromosome;
}

GAChromosomeString GAPopulation::randomSelection()
{
    int populationSize = (int)this->chromosomes.size();
    
    int index = (rand() % populationSize);
    return this->chromosomes.at(index);
}

GAChromosomeString GAPopulation::selectChromosome()
{
    GAChromosomeString chromosome;
    
    switch (this->selectionType)
    {
        case GAPopulationSelectionType::GAPopulationSelectionTypeRouletteWheel: {
            chromosome = this->rouletteWheelSelection();
        }
            break;
            
        case GAPopulationSelectionType::GAPopulationSelectionTypeTournament: {
            chromosome = this->tournamentSelection(this->tournamentSize);
        }
            break;
            
        case GAPopulationSelectionType::GAPopulationSelectionTypeRandom:
        default: {
            chromosome = this->randomSelection();
        }
            break;
    }
    
    return chromosome;
}

#pragma mark -


#pragma mark - Crossover Methods

pair<GAChromosomeString, GAChromosomeString> GAPopulation::onePointCrossover(GAChromosomeString const &chromosomeA, GAChromosomeString const &chromosomeB)
{
    int targetLength = (int)this->targetGenes.length();
    int pivot = (rand() % targetLength);
    
    string genesToCrossOverA = chromosomeA.genes.substr(pivot, targetLength - pivot);
    string genesToCrossOverB = chromosomeB.genes.substr(pivot, targetLength - pivot);
    
    string crossedOverGenesA = chromosomeA.genes.substr(0, pivot) + genesToCrossOverB;
    string crossedOverGenesB = chromosomeB.genes.substr(0, pivot) + genesToCrossOverA;
    
    GAChromosomeString offspringA = GAChromosomeString(crossedOverGenesA);
    GAChromosomeString offspringB = GAChromosomeString(crossedOverGenesB);
    
    float randomMutation = (float)(rand() / (float)RAND_MAX);
    
    if (randomMutation < this->mutationProbability)
    {
        this->mutate(offspringA);
        this->mutate(offspringB);
    }
    
    offspringA.fitness = this->calculateFitness(offspringA);
    offspringB.fitness = this->calculateFitness(offspringB);
    
    return make_pair(offspringA, offspringB);
}

pair<GAChromosomeString, GAChromosomeString> GAPopulation::twoPointCrossover(GAChromosomeString const &chromosomeA, GAChromosomeString const &chromosomeB)
{
    int targetLength = (int)this->targetGenes.length();
    
    int pivotA = (rand() % targetLength);
    int pivotB = (rand() % targetLength);
    
    if (pivotA > pivotB)
    {
        int temp = pivotA;
        pivotA = pivotB;
        pivotB = temp;
    }
    
    string genesToCrossOverA = chromosomeA.genes.substr(pivotA, pivotB - pivotA);
    string genesToCrossOverB = chromosomeB.genes.substr(pivotA, pivotB - pivotA);
    
    string crossedOverGenesA = chromosomeA.genes.substr(0, pivotA) + genesToCrossOverB + chromosomeA.genes.substr(pivotB, targetLength - pivotB);
    string crossedOverGenesB = chromosomeB.genes.substr(0, pivotA) + genesToCrossOverA + chromosomeA.genes.substr(pivotB, targetLength - pivotB);
    
    GAChromosomeString offspringA = GAChromosomeString(crossedOverGenesA);
    GAChromosomeString offspringB = GAChromosomeString(crossedOverGenesB);
    
    float randomMutation = (float)(rand() / (float)RAND_MAX);
    
    if (randomMutation < this->mutationProbability)
    {
        this->mutate(offspringA);
        this->mutate(offspringB);
    }
    
    offspringA.fitness = this->calculateFitness(offspringA);
    offspringB.fitness = this->calculateFitness(offspringB);
    
    return make_pair(offspringA, offspringB);
}

void GAPopulation::crossoverPopulation()
{
    int elitismOffset = 0;
    int populationSize = (int)this->chromosomes.size();
    
    vector<GAChromosomeString> buffer = vector<GAChromosomeString>();
    
    if (this->elitismEnabled)
    {
        elitismOffset = (int)(this->elitismPercentage * populationSize);
        
        sort(this->chromosomes.begin(), this->chromosomes.end());
        
        for (int i = 0; i < elitismOffset; i++)
            buffer.push_back(this->chromosomes.at(i));
    }
    
    for (int i = elitismOffset; i < populationSize; i += 2)
    {
        GAChromosomeString chromosomeA = this->selectChromosome();
        GAChromosomeString chromosomeB = this->selectChromosome();
        
        float randomCrossover = (float)(rand() / (float)RAND_MAX);
        
        if (randomCrossover < this->crossoverProbability)
        {
            pair<GAChromosomeString, GAChromosomeString> offspring;
            
            switch (this->crossoverType)
            {
                case GAPopulationCrossoverType::GAPopulationCrossoverTypeTwoPoint: {
                    offspring = this->twoPointCrossover(chromosomeA, chromosomeB);
                }
                    break;
                    
                case GAPopulationCrossoverType::GAPopulationCrossoverTypeOnePoint:
                default: {
                    offspring = this->onePointCrossover(chromosomeA, chromosomeB);
                }
                    break;
            }
            
            buffer.push_back(offspring.first);
            buffer.push_back(offspring.second);
        }
        else {
            buffer.push_back(chromosomeA);
            buffer.push_back(chromosomeB);
        }
    }
    
    this->chromosomes = vector<GAChromosomeString>(buffer);
}

#pragma mark -


#pragma mark - Fitness Methods

GAChromosomeString GAPopulation::getFittest()
{
    GAChromosomeString bestChromosome = this->chromosomes.at(0);
    
    int populationSize = (int)this->chromosomes.size();
    
    for (int i = 0; i < populationSize; i++)
    {
        GAChromosomeString chromosome = this->chromosomes.at(i);
        
        if (chromosome.fitness < bestChromosome.fitness)
            bestChromosome = chromosome;
    }
    
    return bestChromosome;
}

int GAPopulation::calculateFitness(GAChromosomeString const &theChromosome)
{
    int fitness = 0;
    
    for (int i = 0; i < theChromosome.genes.length(); i++)
        if (theChromosome.genes.at(i) == this->targetGenes.at(i))
            fitness += 1;
    
    return abs(fitness - (int)this->targetGenes.length());
}

#pragma mark -


#pragma mark - Evolution Methods

void GAPopulation::evolve()
{
    int generationCount = 0;
    
    while (this->getFittest().fitness != 0)
    {
        this->crossoverPopulation();
        
        printf("[Generation %d]\n", generationCount);
        printf("Genes: %s\n", this->getFittest().genes.c_str());
        printf("Fitness: %d\n\n", this->getFittest().fitness);
        
        generationCount++;
    }
}

#pragma mark -


#pragma mark - Helper Methods

char GAPopulation::generateRandomGene()
{
    int index = (rand() % kCharacterSet.length());
    return kCharacterSet.at(index);
}

string GAPopulation::generateRandomGenes(int theLength)
{
    string genes = "";
    
    for (int i = 0; i < theLength; i++)
    {
        int index = (rand() % kCharacterSet.length());
        genes += kCharacterSet.at(index);
    }
    
    return genes;
}

#pragma mark -
